torch/_inductor/lowering.py

```diff
@@ -5930,6 +5929,117 @@ def while_loop(cond_fn, body_fn, operands):
     return list(map(TensorBox.create, result))
 
 
+@register_lowering(torch.ops.higher_order.templated_attention)
+def templated_attention(*args, **kwargs):
+    from torch._prims_common import make_contiguous_strides_for
+    from .ir import (
+        ComputedBuffer,
+        FixedLayout,
+        FlexibleLayout,
+        InputBuffer,
+        StorageBox,
+        TensorBox,
+    )
+
+    query, key, value, subgraph = args
+
+    def create_placeholder(name: str, dtype: torch.dtype) -> InputBuffer:
+        return TensorBox.create(
+            InputBuffer(
+                name,
+                FixedLayout(
+                    query.get_device(),
+                    dtype,
+                    [
+                        1,
+                    ],
+                    [
+                        1,
+                    ],
+                ),
+            )
+        )
+
+    scalar_inps = ["score", "b", "h", "m", "n"]
+    env = {}
+    cnt = 0
+    placeholder_inps = [
+        create_placeholder(name, dtype)
+        for name, dtype in [
+            ("score", query.get_dtype()),
+            ("b", torch.int64),
+            ("h", torch.int64),
+            ("m", torch.int64),
+            ("n", torch.int64),
+        ]
+    ]
+    for node in subgraph.graph_module.graph.nodes:
+        # There are two classes of placeholder inpts that we need
+        # to handle differently. For the first n_scalar_inps inputs
+        # we expect that these placeholders were generated by the make_fx call
+        # in the templated Attention HOP. So we need to create a new placeholder
+        # TensorBox for each of these inputs. For the rest of the inputs we
+        # expect that these are lifted inputs that fill up the '*other_buffers'
+        # tuple and already have corresponding TensorBoxes passed in as args.
+        if node.op == "placeholder":
+            is_lifted_input = cnt >= len(scalar_inps)
+            env[node] = args[cnt - 1] if is_lifted_input else placeholder_inps[cnt]
+            cnt += 1
+        elif node.op == "call_function":
+            # For call_function we use the defulat lowerings and pass in the
+            # already created TensorBoxes as args
+            from torch.utils._pytree import tree_map
+
+            env[node] = lowerings[node.target](
```

Comment:
I wonder what happens if we accidentally realize a node during lowering. For example, if they use an op that we can't fuse into.

Reply:
Could maybe use the `lower_pointwise_subgraph` utility from #119430. It overrides `V.graph` so when a lowering tries to register a new buffer or mutation we can raise an error.

Reply:
Yeah right now we get pretty ugly/unhelpful error this sounds like a great follow up
``` Python
w/ 
def bad_mod(score, b, h, m, n):
    return torch.nn.functional.softmax(score, dim=-1)

  File "/home/drisspg/meta/score-mod-api/torch/_inductor/graph.py", line 852, in call_function
    out = lowerings[target](*args, **kwargs)
  File "/home/drisspg/meta/score-mod-api/torch/_inductor/lowering.py", line 300, in wrapped
    out = decomp_fn(*args, **kwargs)
  File "/home/drisspg/meta/score-mod-api/torch/_inductor/lowering.py", line 6063, in templated_attention
    env[node] = lowerings[node.target](
torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
LoweringException: KeyError: <OpOverload(op='aten._softmax', overload='default')>
  target: templated_attention
  args[0]: TensorBox(StorageBox(
    InputBuffer(name='primals_1', layout=FixedLayout('cuda', torch.bfloat16, size=[1, 1, 256, 64], stride=[16384, 16384, 64, 1]))
  ))
  args[1]: TensorBox(StorageBox(
    InputBuffer(name='primals_2', layout=FixedLayout('cuda', torch.bfloat16, size=[1, 1, 256, 64], stride=[16384, 16384, 64, 1]))
  ))
  args[2]: TensorBox(StorageBox(
    InputBuffer(name='primals_3', layout=FixedLayout('cuda', torch.bfloat16, size=[1, 1, 256, 64], stride=[16384, 16384, 64, 1]))
  ))
  args[3]: Subgraph(name='sdpa_score', graph_module=<lambda>(), graph=None)

Set TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information


You can suppress this exception and fall back to eager by setting:
    import torch._dynamo
    torch._dynamo.config.suppress_errors = True
    ```